lib/spdy-transport/connection.js

```diff
@@ -281,7 +281,7 @@ Connection.prototype._handleFrame = function _handleFrame(frame) {
 
   // Session window update
   if (frame.type === 'WINDOW_UPDATE' && frame.id === 0) {
-    if (state.version < 3.1 && state.autoSpdy31) {
```

Comment:
The case I'm working with is in **plain** mode.

If `autoSpdy31` is set to `false`, then the only way for the SPDY server to detect a SPDY/3.1 client is the `id=0 WINDOW_UPDATE`. However, since `autoSpdy31` is set to `false`, the highlighted line above causes the SPDY/3.1 client to be treated as a SPDY/3 client, leading to issues when the amount of data transferred exceeds the INITIAL_WINDOW_SIZE for the connection. With no session-level flow control in SPDY/3, the SPDY/3.1 client stops transmitting after the INITIAL_WINDOW_SIZE bytes have been sent and no recovery is possible because SPDY/3 doesn't deal with `id=0 WINDOW_UPDATE` frames.

On the other hand, when we set `autoSpdy31` to `true`, the SPDY server sends an `id=0 WINDOW_UPDATE` (https://github.com/indutny/spdy-transport/blob/master/lib/spdy-transport/connection.js#L246-L247) when a new client connects (due to the second half of the OR expression). This works fine for SPDY/3.1 clients, but with a SPDY/3 client, the `id=0 WINDOW_UPDATE` results in a `GOAWAY` frame from the client since SPDY/3 doesn't deal with session-level flow control (iSPDY does this because it fails to locate a stream with id=0 and decides to drop the connection: https://github.com/Voxer/iSPDY/blob/master/src/ispdy.m#L1227-L1242).

Thus, it seems that the `autoSpdy31` flag serves as a big hammer that treats all clients as SPDY/3.1 clients when the flag is set, and as SPDY/3 clients when the flag is unset.

What we're looking for is the ability to handle both SPDY/3 and SPDY/3.1 clients in plain mode.

Removing the `autoSpdy31` flag check in the line highlighted above produces the desired result, gracefully upgrading SPDY/3.1 clients, while allowing SPDY/3 clients to proceed unmodified.